src/include/miopen/tensor.hpp

```diff
@@ -175,37 +176,44 @@ struct TensorDescriptor : miopenTensorDescriptor
 
     std::string ToString() const;
 
-    template <class Vector, class Op>
-    static inline std::vector<int64_t> sort_permutation(const Vector& data, Op op)
+    std::string GetLayout(std::string labels) const
```

Comment:
Seek for feedback on this:

- If we do store layout, then get rid of labels because layout will be based on a default set of labels.
- If we don't store layout, then I can bring the `setLayout()` logic to here, and instead of setting it to a class member, return it.

Reply:
If you want to cache the computation you could cache the permutation, and the layout can apply the permutation to the string layout.

Reply:
The disadvantage of storing the permutation or layout string in the tensor is that you can't use the other tensors to help infer the correct layout.

So, alternatively, we could compute the layout in the ProblemDescription. If we write a `IsLayout("NCHW", "NHWC")` function to check if the tensor has the layout. Then you can just loop over the two layouts:

```cpp
std::set<std::string> inputLayouts;
std::set<std::string> outputLayouts;

for(const std::string& layout:{"NCHW", "NHWC"})
{
    if (in.IsLayout("NCHW", layout))
        inputLayouts.insert(layout);
    if (out.IsLayout("NCHW", layout))
        outputLayouts.insert(layout);
}
```

Then you know if the layout is ambiguous because the set may be larger than one element. When its ambiguous, its likely that both input and output are not ambiguous so you can do a set_intersection to prune out the ambiguous layouts. A similar approach can be applied to the filter.

Reply:
> If we write a IsLayout("NCHW", "NHWC") function to check if the tensor has the layout. 

I like this for its elegance. However, I think the problem of this approach is:
- Reduced the supported layout from arbitrary(by igemm kernels) to 2(NCHW and NHWC)
- Other components inside MIOpen can reach at different conclusion than the decision made in `ProblemDescription`. @carlushuang's rest of PR has a number of use cases where he will invoke `GetLayout()` on an individual tensor. At those use cases, `ProblemDescription` is not available in the context. The conclusions made with individual `GetLayout()` can be inconsistent with conclusion made in `ProblemDescription`.

We can adopt this if we agree:
- Do not support other non-trivial layout
- @carlushuang Need to figure out how to get coherent views of layout in different component of MIOpen.

Reply:
> If you want to cache the computation you could cache the permutation, and the layout can apply the permutation to the string layout.

Thank you. I like this better than the current approach. Indeed caching the permutation gives more flexibility for us to assign different labels to the layout.

Reply:
> Reduced the supported layout from arbitrary(by igemm kernels) to 2(NCHW and NHWC)

Well when we assign a layout in the ProblemDescription, it can be an enum of NCHW, NHWC, or Unknown. So if we have a solver that supports arbitrary layouts this enum can be ignored. 

> Other components inside MIOpen can reach at different conclusion than the decision made in ProblemDescription

Yes but to be able to do that in the solvers the ProblemDescription needs to store the tensor descriptors.

> rest of PR has a number of use cases where he will invoke GetLayout() on an individual tensor. 

First, in this PR `GetLayout` is only used to compute the output tensor which is only used for testing(and the driver). In MIGraphX, we have a [`with_lens`](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/data-layout/src/shape.cpp#L225) function that will try to propagate the layout, you should add a similar method. 

Second, for testing we should probably explicitly call the output layout we want rather than trying to infer it from the strides of the input. Also, because we will want to test having different input and output layouts in the future possibly.

Thirdly, if we cache a single interpretation of the permutation in the tensor descriptor then there is no room to interpret it as a different layout in ambiguous cases.

Reply:
Yes currently in this PR if all layout is explicitly specified by `--in_layout`, `--fil_layout`, `--out_layout`, will use this layout specification.

Reply:
> At those use cases, ProblemDescription is not available in the context. The conclusions made with individual GetLayout() can be inconsistent with conclusion made in ProblemDescription

Yes, and in ctest we may have not yet created `ProblemDescription` but do need layout information just from `TensorDescriptor`.

Reply:
> So if we have a solver that supports arbitrary layouts this enum can be ignored.

We don't have this enum right now. Are you suggesting to start with only `NCHW` and `NHWC` and come up with the enum in a later API change, then update the heuristic again? 
 - Follow up: Let me rephrase: when it is a `UNKNOWN` enum and and it is ambiguous, we skip the `IsLayout()` checks and directly use the layout getter from `tensor.hpp`.

If this is the case, I see a two step process here, and we don't necessarily address all concerns in this PR alone.
- step 1: `tensor.hpp::IsLayout()` + Find Heuristic in ProblemDescription, only enable `NCHW` + `NHWC` + enforce ctest to know all layout deterministically from user input
- step 2: `tensor.hpp::find_permutation` + API function to set enum layout, enable `nonNCHW/NHWC`

Reply:
I coded step 1 in commit 31e693f.

This commit is independent from 2b80369. It implements the heuristic init as @pfultz2 suggested. We can revert 2b80369 until we reach at a consensus, then implement step 2.